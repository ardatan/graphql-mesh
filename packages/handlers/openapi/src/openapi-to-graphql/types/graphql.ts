// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
