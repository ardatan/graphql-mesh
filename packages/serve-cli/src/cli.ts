import '@graphql-mesh/include/register-tsconfig-paths';
import 'dotenv/config'; // inject dotenv options to process.env
import 'json-bigint-patch'; // JSON.parse/stringify with bigints support

import cluster from 'node:cluster';
import { availableParallelism, release } from 'node:os';
import parseDuration from 'parse-duration';
import {
  Command,
  InvalidArgumentError,
  Option,
  type OptionValues,
} from '@commander-js/extra-typings';
import type {
  MeshServeConfigProxy,
  MeshServeConfigSubgraph,
  MeshServeConfigSupergraph,
} from '@graphql-mesh/serve-runtime';
import type { Logger } from '@graphql-mesh/types';
import { DefaultLogger } from '@graphql-mesh/utils';
import { addCommands } from './commands/index.js';
import { defaultConfigPaths } from './config.js';
import type { ServerConfig } from './server';

export type MeshServeCLIConfig = (
  | MeshServeCLISupergraphConfig
  | MeshServeCLISubgraphConfig
  | MeshServeCLIProxyConfig
) &
  ServerConfig & {
    /**
     * Count of workers to spawn. Defaults to `os.availableParallelism()` when NODE_ENV
     * is "production", otherwise only one (the main) worker.
     */
    fork?: number;
    /**
     * GraphQL schema polling interval in milliseconds.
     *
     * @default 10_000
     */
    pollingInterval?: number;
  };

export interface MeshServeCLISupergraphConfig
  extends Omit<MeshServeConfigSupergraph, 'supergraph'> {
  /**
   * SDL, path or an URL to the Federation Supergraph.
   *
   * Alternatively, CDN options for pulling a remote Federation Supergraph.
   *
   * @default 'supergraph.graphql'
   */
  // default matches commands/supergraph.ts
  supergraph?: MeshServeConfigSupergraph['supergraph'];
}

export interface MeshServeCLISubgraphConfig extends Omit<MeshServeConfigSubgraph, 'subgraph'> {
  /**
   * SDL, path or an URL to the Federation Supergraph.
   *
   * Alternatively, CDN options for pulling a remote Federation Supergraph.
   *
   * @default 'subgraph.graphql'
   */
  // default matches commands/subgraph.ts
  subgraph?: MeshServeConfigSubgraph['subgraph'];
}

export interface MeshServeCLIProxyConfig extends Omit<MeshServeConfigProxy, 'proxy'> {
  /**
   * HTTP executor to proxy all incoming requests to another HTTP endpoint.
   */
  proxy?: MeshServeConfigProxy['proxy'];
}

/**
 * Type helper for defining the config.
 */
export function defineConfig(config: MeshServeCLIConfig) {
  return config;
}

/** The context of the running program. */
export interface CLIContext {
  /** @default new DefaultLogger() */
  log: Logger;
  /** @default 'Mesh Serve' */
  productName: string;
  /** @default 'serve GraphQL federated architecture for any API service(s)' */
  productDescription: string;
  /** @default 'mesh-serve' */
  binName: string;
  /** @default globalThis.__VERSION__ */
  version: string;
}

/** Inferred program options from the root command {@link cli}. */
export type CLIGlobals = CLI extends Command<any, infer O> ? O : never;

export const defaultFork = process.env.NODE_ENV === 'production' ? availableParallelism() : 1;

export type CLI = typeof cli;

export type AddCommand = (ctx: CLIContext, cli: CLI) => void;

/** The root cli of serve-cli. */
let cli = new Command()
  .configureHelp({
    // will print help of global options for each command
    showGlobalOptions: true,
  })
  .addOption(
    new Option(
      '--fork <count>',
      'count of workers to spawn. defaults to `os.availableParallelism()` when NODE_ENV is "production", otherwise only one (the main) worker',
    )
      .env('FORK')
      .argParser(v => {
        const count = parseInt(v);
        if (isNaN(count)) {
          throw new InvalidArgumentError('not a number.');
        }
        return count;
      })
      .default(defaultFork),
  )
  .addOption(
    new Option(
      '-c, --config-path <path>',
      `path to the configuration file. defaults to the following files respectively in the current working directory: ${defaultConfigPaths.join(', ')}`,
    ).env('CONFIG_PATH'),
  )
  .option(
    '-h, --host <hostname>',
    'host to use for serving',
    release().toLowerCase().includes('microsoft') ? '127.0.0.1' : '0.0.0.0',
  )
  .addOption(
    new Option('-p, --port <number>', 'port to use for serving')
      .env('PORT')
      .default(4000)
      .argParser(v => {
        const port = parseInt(v);
        if (isNaN(port)) {
          throw new InvalidArgumentError('not a number.');
        }
        return port;
      }),
  )
  .addOption(
    new Option('--polling <duration>', 'schema polling interval in human readable duration')
      .env('POLLING')
      .default(10_000, '10s')
      .argParser(v => {
        const interval = parseDuration(v);
        if (!interval) {
          throw new InvalidArgumentError('not a duration.');
        }
        return interval;
      }),
  )
  .option('--no-masked-errors', "don't mask unexpected errors in responses")
  .option(
    '--masked-errors',
    'mask unexpected errors in responses (default: true)',
    // we use "null" intentionally so that we know when the user provided the flag vs when not
    // see here https://github.com/tj/commander.js/blob/970ecae402b253de691e6a9066fea22f38fe7431/lib/command.js#L655
    null,
  )
  .addOption(
    new Option(
      '--hive-registry-token <token>',
      'Hive registry token for usage metrics reporting',
    ).env('HIVE_REGISTRY_TOKEN'),
  );

// @inject-version globalThis.__VERSION__ here

export function run(userCtx: Partial<CLIContext>) {
  const ctx: CLIContext = {
    log: new DefaultLogger(),
    productName: 'Mesh',
    productDescription: 'serve GraphQL federated architecture for any API service(s)',
    binName: 'mesh-serve',
    version: globalThis.__VERSION__,
    ...userCtx,
  };

  const { binName, productDescription, version } = ctx;
  cli = cli.name(binName).description(productDescription);
  cli.version(version);

  const log = ctx.log.child(
    cluster.worker?.id ? `${ctx.productName} worker #${cluster.worker.id}` : ctx.productName,
  );

  addCommands({ ...ctx, log }, cli);

  return cli.parseAsync();
}
