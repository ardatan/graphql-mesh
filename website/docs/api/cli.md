---
id: cli
title: CLI
sidebar_label: CLI
---

WIP!