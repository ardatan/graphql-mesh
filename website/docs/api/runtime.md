---
id: runtime
title: Runtime
sidebar_label: Runtime
---

WIP!