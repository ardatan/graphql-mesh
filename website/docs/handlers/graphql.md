---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

This handler allows you to load remote GraphQL schemas and use it with schema-stitching, based on `graphql-tools-fork`.

To get started, install the handler library from NPM:

```
$ yarn add @graphql-mesh/graphql
```

Now, you can use it directly in your Mesh config file:

```yml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

## Config API Reference

{@import ../generated-markdown/GraphQLHandler.generated.md}
