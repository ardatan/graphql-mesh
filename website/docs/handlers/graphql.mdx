---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageInstall packages="@graphql-mesh/graphql" />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

> You can check out our example that uses schema stitching with a PostgreSQL data source.
> [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

<p>&nbsp;</p>

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

<p>&nbsp;</p>

### From Environmental Variable

`MY_API_TOKEN` is the name of the environmental variable you have the value.

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Config API Reference

{@import ../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md}
or
{@import ../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md}
