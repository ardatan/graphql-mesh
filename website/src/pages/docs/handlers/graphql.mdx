import { PackageCmd, Callout } from '@theguild/components'

# GraphQL

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageCmd packages={['@graphql-mesh/graphql']} />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

<Callout>
  You can check out our example that uses schema stitching with a PostgreSQL data source.
  [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

</Callout>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          # Use "{context.headers['x-my-api-token']}" if you want just the value of the header
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

### From Environment Variables

`MY_API_TOKEN` is the name of the environment variable you have the value.

```yaml {7}
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

## Local Schemas

We recommend providing local schema by using the `additionalTypeDefs` and `additionalResolvers` configuration options.

However, it is also possible to use a local GraphQL Schema instance as a GraphQL Mesh source, as showcased below:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        schema: ./my-local-schema.ts
```

```ts filename="my-local-schema.ts"
import { makeExecutableSchema } from '@graphql-tools/schema'
export default makeExecutableSchema({
  typeDefs: /* GraphQL */ `
    type Query {
      foo: String
    }
  `,
  resolvers: {
    Query: {
      foo: () => 'FOO'
    }
  }
})
```

## Fetch Strategies and Multiple HTTP endpoints for the same source

If you want to have an advanced fetch strategy for the GraphQL source such as retrying twice or timeout in 30 seconds etc.
Also, you can have different HTTP endpoints for a single source, and you can configure Mesh to get a better execution flow.

For example, you can make a request to both endpoints and return the fastest response with `race` strategy.

All `fetch` strategies can be combined to create the ultimate execution flow:

<details>
 <summary>`retry`</summary>

The `retry` mechanism allow you to specify the retry attempts for a single GraphQL endpoint/source.

The retry flow will execute in both conditions: a network error, or due to a runtime error.

```yaml {6}
sources:
  - name: uniswapv2
    handler:
      graphql:
        endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
        retry: 2 # specify here, if you have an unstable/error prone indexer
```

</details>

<details>
 <summary>`timeout`</summary>

The `timeout` mechanism allow you to specify the `timeout` for a given GraphQL endpoint.

```yaml {6}
sources:
  - name: uniswapv2
    handler:
      graphql:
        endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
        timeout: 5000 # 5 seconds
```

</details>

<details>
 <summary>`fallback`</summary>

The `fallback` mechanism allow you to specify use more than one GraphQL endpoint, for the same source.

This is helpful if you have a fallback endpoint for the same GraphQL API.

```yaml {5}
sources:
  - name: uniswapv2
    handler:
      graphql:
        strategy: fallback
        sources:
          - endpoint: https://bad-uniswap-v2-api.com
            retry: 2
            timeout: 5000
          - endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
```

</details>

<details>
 <summary>`race`</summary>

The `race` mechanism allow you to specify use more than one GraphQL endpoint, for the same source, and race on every execution.

If you have different places that service is deployed, this is useful to get the fastest response by racing them.

```yaml {5}
sources:
  - name: uniswapv2
    handler:
      graphql:
        strategy: race
        sources:
          - endpoint: https://bad-uniswap-v2-api.com
          - endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
```

</details>

## Config API Reference

import API1 from '../../../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md'
import API2 from '../../../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md'

<API1 />
or
<API2 />
