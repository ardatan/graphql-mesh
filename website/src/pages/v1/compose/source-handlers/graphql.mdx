---
description: GraphQL Mesh can load remote GraphQL schemas, fetch from a CDN or registry
---

import { Callout } from '@theguild/components'

# GraphQL

<Callout>
  This page is currently under construction and expected to change. Please feel free to reach out to
  us directly in case you are having any troubles.
</Callout>

This handler allows you to load remote GraphQL schemas and compose them into a supergraph.

## How to use?

If you want to load a remote GraphQL subgraph by introspecting the endpoint, you can only provide
`endpoint` in the configuration like below;

```ts filename="mesh-config.ts"
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Users', {
        endpoint: 'http://localhost:4001/users'
      })
    }
  ]
})
```

### Load the schema via URL or local file (Hive CDN etc)

If you want to load a remote GraphQL subgraph from a CDN, you can provide `source` in the
configuration like below;

In this case, Mesh will take the schema from `source`, but execute the operations against the
`endpoint`.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('GitHub', {
        endpoint: 'https://api.github.com/graphql',
        // You can provide your SDL or introspection seperately
        source: 'https://docs.github.com/public/schema.docs.graphql',

        operationHeaders: {
          // This forwards the header from the incoming request to the remote server
          authorization: 'Bearer {context.headers["x-my-api-token"]}'
        }
      })
    }
  ]
})
```

## Headers

[Read about configuration and examples](/v1/compose/source-handlers#setting-headers)

## Fetch Strategies and Multiple HTTP endpoints for the same source

If you want to have an advanced fetch strategy for the GraphQL source such as retrying twice or
timeout in 30 seconds etc. Also, you can have different HTTP endpoints for a single source, and you
can configure Mesh to get a better execution flow.

For example, you can make a request to both endpoints and return the fastest response with `race`
strategy.

All `fetch` strategies can be combined to create the ultimate execution flow:

<details>
 <summary>`retry`</summary>

The `retry` mechanism allow you to specify the retry attempts for a single GraphQL endpoint/source.

The retry flow will execute in both conditions: a network error, or due to a runtime error.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2',
        // Specify here, if you have an unstable/error prone indexer
        retry: 2
      })
    }
  ]
})
```

</details>

<details>
 <summary>`timeout`</summary>

The `timeout` mechanism allow you to specify the `timeout` for a given GraphQL endpoint.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2',
        // 5 Seconds
        timeout: 5_000
      })
    }
  ]
})
```

</details>

<details>
 <summary>`fallback`</summary>

The `fallback` mechanism allow you to specify use more than one GraphQL endpoint, for the same
source.

This is helpful if you have a fallback endpoint for the same GraphQL API.

```ts filename="mesh-config.ts" {7-17}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        strategy: 'fallback',
        sources: [
          {
            endpoint: 'https://bad-uniswap-v2-api.com',
            retry: 2,
            timeout: 5_000
          },
          {
            endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2'
          }
        ]
      })
    }
  ]
})
```

</details>

<details>
 <summary>`race`</summary>

The `race` mechanism allow you to specify use more than one GraphQL endpoint, for the same source,
and race on every execution.

If you have different places that service is deployed, this is useful to get the fastest response by
racing them.

```ts filename="mesh-config.ts" {7-15}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        strategy: 'race',
        sources: [
          {
            endpoint: 'https://bad-uniswap-v2-api.com'
          },
          {
            endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2'
          }
        ]
      })
    }
  ]
})
```

</details>
