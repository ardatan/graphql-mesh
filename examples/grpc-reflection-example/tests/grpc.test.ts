import 'json-bigint-patch';
import { findAndParseConfig } from '@graphql-mesh/cli';
import { getMesh } from '@graphql-mesh/runtime';
import { join } from 'path';

import { printSchema, lexicographicSortSchema } from 'graphql';
import { readFile } from 'fs-extra';

const config$ = findAndParseConfig({
  dir: join(__dirname, '..'),
});
const mesh$ = config$.then(config => getMesh(config));
import startGrpcServer from '../start-server.js';
const grpc$ = startGrpcServer(300);
jest.setTimeout(15000);

describe('gRPC Example', () => {
  it('should generate correct schema', async () => {
    const { schema } = await mesh$;
    expect(printSchema(lexicographicSortSchema(schema))).toMatchSnapshot('grpc-schema');
  });
  it('should get greeting correctly', async () => {
    const GreetingQuery = await readFile(
      join(__dirname, '../example-queries/GetGreeting.query.graphql'),
      'utf8',
    );
    const { execute } = await mesh$;
    await grpc$;
    const result = await execute(GreetingQuery, undefined);
    expect(result).toMatchSnapshot('greeting-result');
  });
  afterAll(() => {
    mesh$.then(mesh => mesh.destroy());
    grpc$.then(grpc => grpc.forceShutdown());
  });
});
