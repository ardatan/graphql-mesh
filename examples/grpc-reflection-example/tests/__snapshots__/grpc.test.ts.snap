// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gRPC Example should generate correct schema: grpc-schema 1`] = `
"schema {
  query: Query
}

directive @grpcMethod(rootJsonName: String, objPath: String, methodName: String, responseStream: Boolean) on FIELD_DEFINITION

directive @grpcConnectivityState(rootJsonName: String, objPath: String) on FIELD_DEFINITION

directive @grpcRootJson(name: String, rootJson: ObjMap, loadOptions: ObjMap) repeatable on OBJECT

type Query @grpcRootJson(name: "Root0", rootJson: "{\\"nested\\":{\\"GreetingRequest\\":{\\"fields\\":{\\"name\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"GreetingResult\\":{\\"fields\\":{\\"greeting\\":{\\"type\\":\\"string\\",\\"id\\":1}},\\"comment\\":null},\\"GreetingService\\":{\\"methods\\":{\\"GetGreeting\\":{\\"requestType\\":\\".GreetingRequest\\",\\"responseType\\":\\".GreetingResult\\"}},\\"comment\\":null}}}") {
  GreetingService_GetGreeting(input: GreetingRequest_Input): GreetingResult @grpcMethod(rootJsonName: "Root0", objPath: "GreetingService", methodName: "GetGreeting", responseStream: false)
  GreetingService_connectivityState(tryToConnect: Boolean): ConnectivityState @grpcConnectivityState(rootJsonName: "Root0", objPath: "GreetingService")
}

type GreetingResult {
  greeting: String
}

input GreetingRequest_Input {
  name: String
}

enum ConnectivityState {
  IDLE
  CONNECTING
  READY
  TRANSIENT_FAILURE
  SHUTDOWN
}

scalar ObjMap"
`;

exports[`gRPC Example should get greeting correctly: greeting-result 1`] = `
{
  "data": {
    "GreetingService_GetGreeting": {
      "greeting": "Hello World",
    },
  },
}
`;
